runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -870,16 +873,26 @@ private Debuggee registerDebuggee(CloudDebugger debuggerClient, String uniquifie
         } catch (InvalidProtocolBufferException e) {
           throw new RuntimeException("Error parsing artifact file payload.", e);
         }
-        if (!BeamUrns.getUrn(RunnerApi.StandardArtifacts.Roles.STAGING_TO)
+        String stagedName;
+        if (BeamUrns.getUrn(RunnerApi.StandardArtifacts.Roles.STAGING_TO)
             .equals(info.getRoleUrn())) {
-          throw new RuntimeException(
-              String.format("unsupported artifact role %s", info.getRoleUrn()));
-        }
-        RunnerApi.ArtifactStagingToRolePayload stagingPayload;
-        try {
-          stagingPayload = RunnerApi.ArtifactStagingToRolePayload.parseFrom(info.getRolePayload());
-        } catch (InvalidProtocolBufferException e) {
-          throw new RuntimeException("Error parsing artifact staging_to role payload.", e);
+          try {
+            RunnerApi.ArtifactStagingToRolePayload stagingPayload =
+                RunnerApi.ArtifactStagingToRolePayload.parseFrom(info.getRolePayload());
+            stagedName = stagingPayload.getStagedName();
+          } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException("Error parsing artifact staging_to role payload.", e);
+          }
+        } else {
+          try {
+            File source = new File(filePayload.getPath());
+            HashCode hashCode = Files.asByteSource(source).hash(Hashing.sha256());
```

Comment:
Do you know how expensive this is ? Could this add a significant overhead to the Job submission time ?

Reply:
Is it possible to use a hash of the File path instead of file data ?

Reply:
I think this won't increase the job submission time since it only covers exceptional cases. In principle, all Java artifacts for Dataflow runner should have staging_to locations however some Python system artifacts like `pickled_main` don't have staging_to information so they need a special care in Java SDK (Python SDK handles them properly but not Java SDK).

Reply:
It might be little confusing if we use a hash of the file path instead of the file data since we already generate staging names from the file data by [default](https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java#L333). I think it's not ideal if two types of staging names (from the file data, from the file name) coexist.

Reply:
Please add tests to make sure that we set "staging_to" in the default path if we don't have such tests already.

Reply:
Done.